### DISCUSSION TITLE
How to use subproccess in a plugin

### DISCUSSION THREAD

**Discussion by deafwired** [nicotine-plus/nicotine-plus#2596] [Q&A] [answered]
To give some information on what my plugin is doing; once a music file is done downloading, I use ffprobe through subprocess.run to get the name and artist name. I then use curl to get that song's lyrics on genius and after that I use ffmpeg to add it to the file. But every time I use subprocess.run (at least on windows), it opens a new terminal, loads, then immediately deletes the process. I want to have it run a command without opening a new terminal, taking the users focus from whatever window they are in, as it could be annoying having it happen 3 times for every song they download.

## TL:DR
how do I get subprocess to not make a new window, or is there another solution?

**Comment by slook** [comment]
> I use ffprobe through subprocess.run to get the name and artist name

You can use the `TinyTag()` class to extract `artist` and `title` metadata tags from local files.

For example the search.py , shares.py and userbrowse.py modules get the `bitrate` and `duration` values this way. See the [upstream documentation](https://github.com/devsnd/tinytag#readme) for more details.

TinyTag cannot get data from the internet nor write tags however.

> it opens a new terminal

See related issue https://github.com/nicotine-plus/nicotine-plus/issues/1847#issuecomment-1019573694

**Reply by deafwired** [reply]
The TinyTag class is useful for getting rid of one of the subprocess' but i am trying to find a solution that doesnt need the user to install something, thanks for the help though.

**Reply by slook** [reply]
Perhaps look at calling the `execute_command()` function in [utils.py](https://github.com/nicotine-plus/nicotine-plus/blob/7951aded263daa074da437054f95ea0533c84b1b/pynicotine/utils.py#L207-L315).

If we can close #1847 then your problem might be solvable.

**Reply by deafwired** [reply]
the [utils.py](https://github.com/nicotine-plus/nicotine-plus/blob/7951aded263daa074da437054f95ea0533c84b1b/pynicotine/utils.py#L207-L315) uses subprocess.Popen, and although as far as i could see Popen does not return its output, I used urllib to get the genius lyrics (I got 403 because there was no useragent) and Popen to use ffmpeg. however when looking at the tinytag class the nicotine tinytag seems to be different from the other one and gave me an error
![image](https://github.com/nicotine-plus/nicotine-plus/assets/37221707/97e21681-bf91-4c7b-a4ea-e61f68f9d30a)

**Reply by slook** [reply]
@mathiascode whats happening with the required args in tinytag here? Is there a reason why the package uses a different syntax in our version than the one from upstream?

**Reply by mathiascode** [reply]
Does `execute_command(command, returnoutput=True)` work? I made a change in 7b6e436106859df37d3289fd2fd789e6db62de38 that hides the console window. If you want to use subprocess directly, you could add something similar to your plugin.

**Reply by mathiascode** [reply]
> @mathiascode whats happening with the required args in tinytag here? Is there a reason why the package uses a different syntax in our version than the one from upstream?

The size argument was added to improve performance, since we already know the size, but I could make it optional to avoid breaking syntax.

**Reply by deafwired** [reply]
i think its fine i was just confused by the different parameters between the two TinyTag's, let me try implementing it with the size

**Reply by deafwired** [reply]
let me know what i'm doing wrong because i'm confused about the differences between this tinytag. I just wanna get the album artist and track name
```python
from pynicotine.external.tinytag import TinyTag
import os


class metadata:
    def __init__(self, file):
        tags = TinyTag.get(file, os.path.basename(file), os.path.getsize(file))
        return tags.artist, tags.title, file.album

```

**Reply by slook** [reply]
`file.album` -> `tags.album`

Since https://github.com/nicotine-plus/nicotine-plus/commit/cadc221710597a3a1d55fb1639b067938fb718a8 you do not need to know the file size to make a tinytag call anymore. Afaik it should be the same to use as the upstream version now.

**Reply by deafwired** [reply]
that was incorrect but i could not even get to that part of the code, if i try to implement the example code from the upstream version
```python
# testing plugin
from pynicotine.pluginsystem import BasePlugin
from pynicotine.external.tinytag import TinyTag


class Plugin(BasePlugin):
    def __init__(self):
        super().__init__()
        path = "C:/VeryRealMusic/Artist/Album/01 - Song.mp3"
        tags = TinyTag.get(path)
        self.log(tags.artist, tags.title, tags.album)
```
it would give me an error about `TypeError: TinyTag.get() missing 1 required positional argument: 'filename'`
I then tried to do 
```python
# testing plugin
from pynicotine.pluginsystem import BasePlugin
from pynicotine.external.tinytag import TinyTag


class Plugin(BasePlugin):
    def __init__(self):
        super().__init__()
        path = "C:/VeryRealMusic/Artist/Album/01 - Song.mp3"
        tags = TinyTag()
        tags = tags.get(path)
        self.log(tags.artist, tags.title, tags.album)
```
but get `File "D:/a/nicotine-plus/nicotine-plus/pynicotine/external/tinytag.py", line 138, in get 
TypeError: tuple for endswith must only contain str, not bytes`
let me know if i am not understanding how to use this

**Reply by slook** [reply]
It doesn't look like your fault. @mathiascode needs to figure out if it is possible to implement the performance optimizations we need for the core while maintaining the correct syntax which is expected for calling `TinyTag()`.

Otherwise, it expects both the full path (including filename) and the filename (without foldername), which can be obtained with: 

`filename = path.rsplit("\", maxsplit=1)`

Ideally there wouldn't be any modifications to the upstream module whatsoever. TinyTag may have a variety of uses for plugin developers.

**Reply by mathiascode** [reply, marked as answer]
For now, this should work (use encode_path in your second code example if this one doesn't work):
```
from pynicotine.external.tinytag import TinyTag
from pynicotine.utils import encode_path

tag = TinyTag.get(encode_path(path))
```

I'll spend some time syncing our tinytag.py with the upstream project soon.

> Otherwise, it expects both the full path (including filename) and the filename (without foldername), which can be obtained with:

Not sure where you saw this, but it shouldn't be the case.

**Reply by deafwired** [reply]
That worked, I appreciate the effort you're putting into nicotine+ its a great app
